### PR TITLE
Variable Door Delay + Timer Based Animations

### PR DESCRIPTION
**Pull request by LemonInTheDark**
### Variable Door Delay

Door opening/closing delay times are currently static, but many doors do not have animations that FIT the actual timings.

It would be better if subtypes cound declare how long each animation takes, and how long it takes for opening to say, become passable and such.

Let's do that.

### Timer Based Animations

Currently all doors use flick() to do their animations. This is fine right NOW, but fucks with walleniong because we have to split most things into segments to make layering work.

So rather then flick let's use client timers and update_icon_state to achive our effects, alongside a proc that lets us do other effects (like sound) on playing an animation

## Why It's Good For The Game

Door behavior and visuals better match up, wallening compatability upstreaming.

## Changelog
:cl:
add: Most door animations now better line up with when they are/are not passable.
/:cl:

### CODE REVIEW

File: code/game/machinery/doors/door.dm [hunk 1]
```diff
@@ -405,24 +407,67 @@
...
+/// Returns the delay to use for the passed in animation
+/// We'll do our cleanup once the delay runs out
+/obj/machinery/door/proc/animation_length(animation)
+	switch(animation)
+		if("opening")
+			return 0.6 SECONDS
+		if("closing")
+			return 0.6 SECONDS
+		if("deny")
+			return 0.3 SECONDS
```
Comment (MrMelbert) [inline code/game/machinery/doors/door.dm:442]: These should probably be defines yeah?

Comment (LemonInTheDark) [reply code/game/machinery/doors/door.dm]: u right, handled.

File: code/datums/wires/airlock.dm [hunk 2]
```diff
@@ -126,7 +126,7 @@
 			A.update_appearance()
 		if(WIRE_IDSCAN) // Pulse to disable emergency access and flash the red lights.
 			if(A.hasPower() && A.density)
-				A.do_animate("deny")
+				A.run_animation("deny")
```
Comment (MrMelbert) [inline code/datums/wires/airlock.dm:129]:
```suggestion
				A.run_animation(DOOR_DENY_ANIMATION)
```

File: code/game/machinery/doors/airlock.dm [hunk 3]
```diff
@@ -1312,22 +1337,25 @@
 			filler.density = TRUE
 		flags_1 |= PREVENT_CLICK_UNDER_1
 		air_update_turf(TRUE, TRUE)
-	sleep(0.1 SECONDS)
+	var/unpassable_delay = animation_segment_delay("AIRLOCK_CLOSING_UNPASSABLE")
```
Comment (MrMelbert) [inline code/game/machinery/doors/airlock.dm:1340]: Find and Replace moment
```suggestion
	var/unpassable_delay = animation_segment_delay(AIRLOCK_CLOSING_UNPASSABLE)
```

File: code/game/machinery/doors/airlock.dm [hunk 4]
```diff
@@ -1312,22 +1337,25 @@
...
+	sleep(unpassable_delay)
 	if(!air_tight)
 		set_density(TRUE)
 		if(multi_tile)
 			filler.density = TRUE
 		flags_1 |= PREVENT_CLICK_UNDER_1
 		air_update_turf(TRUE, TRUE)
-	sleep(0.4 SECONDS)
+	var/opaque_delay = animation_segment_delay("AIRLOCK_CLOSING_OPAQUE") - unpassable_delay
```
Comment (MrMelbert) [inline code/game/machinery/doors/airlock.dm:1348]:
```suggestion
	var/opaque_delay = animation_segment_delay(AIRLOCK_CLOSING_OPAQUE) - unpassable_delay
```

File: code/game/machinery/doors/airlock.dm [hunk 5]
```diff
@@ -1312,22 +1337,25 @@
...
+	sleep(opaque_delay)
 	if(dangerous_close)
 		crush()
 	if(visible && !glass)
 		set_opacity(TRUE)
 		if(multi_tile)
 			filler.set_opacity(TRUE)
 	update_freelook_sight()
-	sleep(0.1 SECONDS)
+	var/close_delay = animation_segment_delay("AIRLOCK_CLOSING_FINISHED") - unpassable_delay - opaque_delay
```
Comment (MrMelbert) [inline code/game/machinery/doors/airlock.dm:1357]:
```suggestion
	var/close_delay = animation_segment_delay(AIRLOCK_CLOSING_FINISHED) - unpassable_delay - opaque_delay
```

File: code/game/objects/items/puzzle_pieces.dm [hunk 6]
```diff
@@ -86,6 +86,18 @@
...
+	switch(animation)
+		if(DOOR_OPENING_ANIMATION)
+			return 1.0 SECONDS
+
+/obj/machinery/door/puzzle/animation_segment_delay(animation)
+	switch(animation)
+		if("opening_passable")
+			return 0.8 SECONDS
+		if("opening_done")
+			return 1.0 SECONDS
```
Comment (MrMelbert) [inline code/game/objects/items/puzzle_pieces.dm:99]: Missed one, also `"opening_done"` seems unused?

File: code/modules/mapfluff/ruins/icemoonruin_code/library.dm [hunk 7]
```diff
@@ -5,6 +5,18 @@
...
+	switch(animation)
+		if(DOOR_OPENING_ANIMATION)
+			return 1.2 SECONDS
+
+/obj/machinery/door/puzzle/keycard/library/animation_segment_delay(animation)
+	switch(animation)
+		if("opening_passable")
+			return 1.0 SECONDS
+		if("opening_done")
+			return 1.2 SECONDS
```
Comment (MrMelbert) [inline code/modules/mapfluff/ruins/icemoonruin_code/library.dm:18]: Ok missed two

File: code/game/objects/items/puzzle_pieces.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/game/objects/items/puzzle_pieces.dm:99]: renamed to the finished suffix

File: code/game/machinery/doors/airlock.dm [hunk 8]
```diff
@@ -607,6 +607,28 @@
...
+		if(AIRLOCK_OPENING_PASSABLE)
+			return 0.5 SECONDS
+		if(AIRLOCK_OPENING_FINISHED)
+			return 0.6 SECONDS
+		if("AIRLOCK_CLOSING_UNPASSABLE")
+			return 0.2 SECONDS
+		if("AIRLOCK_CLOSING_OPAQUE")
+			return 0.5 SECONDS
+		if("AIRLOCK_CLOSING_FINISHED")
+			return 0.6 SECONDS
```
Comment (san7890) [inline code/game/machinery/doors/airlock.dm:630]: you're using the strings and not the defines here.

**Review by san7890 — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/airlock.dm (hunk 8 above)
Comment (LemonInTheDark) [reply code/game/machinery/doors/airlock.dm:630]: I thought I got em all, fug

**Review by Ghommie — APPROVED** [review]

**Pull request merged**